include/mbgl/style/expression/match.hpp

```diff
@@ -0,0 +1,277 @@
+            result = {value->template get<std::string>()};
+        } else {
+            return CompileError {
+                "Match inputs must be either literal integer or string values or arrays of integer or string values.",
+                key
+            };
+        }
+        
+        if (!inputType) {
+            inputType = type;
+        } else {
+            const auto& expected = toString(*inputType);
+            const auto& actual = toString(*type);
+            // TODO: replace with real == check
+            if (expected != actual) {
+                return CompileError {
+                    "Expected " + expected + " but found " + actual + " instead.",
+                    key
+                };
+            }
+        }
+        
+        return *result;
+    }
+    
+    TypecheckResult typecheck(std::vector<CompileError>& errors) const override {
+        auto checkedInput = input->typecheck(errors);
+        if (!checkedInput) {
+            return TypecheckResult();
+        }
+        auto checkedOtherwise = otherwise->typecheck(errors);
+        if (!checkedOtherwise) {
+            return TypecheckResult();
+        }
+
+        auto error = matchType(inputType, (*checkedInput)->getType());
+        if (error) {
+            errors.push_back({ *error, input->getKey() });
+        }
+        
+        optional<type::Type> outputType;
+        std::vector<std::pair<MatchKey, std::unique_ptr<TypedExpression>>> checkedCases;
+        for (const auto& pair : cases) {
+            auto checkedOutput = pair.second->typecheck(errors);
+            if (!checkedOutput) continue;
+            if (!outputType) {
+                outputType = (*checkedOutput)->getType();
+            } else {
+                error = matchType(*outputType, (*checkedOutput)->getType());
+                if (error) {
+                    errors.push_back({ *error, pair.second->getKey() });
+                    continue;
+                }
+            }
+            checkedCases.emplace_back(pair.first, std::move(*checkedOutput));
+        }
+        
+        error = matchType(*outputType, (*checkedOtherwise)->getType());
+        if (error) {
+            errors.push_back({ *error, otherwise->getKey() });
+        }
+        
+        if (inputType.is<type::StringType>()) {
+            return TypecheckResult(std::make_unique<TypedMatch<std::string>>(
+                std::move(*checkedInput), std::move(checkedCases), std::move(*checkedOtherwise)
+            ));
+        } else if (inputType.is<type::NumberType>()) {
+            return TypecheckResult(std::make_unique<TypedMatch<int64_t>>(
+                std::move(*checkedInput), std::move(checkedCases), std::move(*checkedOtherwise)
+            ));
+        }
+        
+        assert(false);
+        return TypecheckResult();
+    }
+    
+private:
+    static bool isIntegerValue(const mbgl::Value& v) {
+        return v.match(
+            [] (uint64_t) { return true; },
```

Comment:
Need to handle overflowing of `int64_t` for this case. Also, what about overflowing `Number.MAX_SAFE_INTEGER` in JS? Should we enforce a range that's compatible with JS?